include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+        : Knob(displayName, flagName, flagDesc)
+    {
+        if (choices.size() < 1) {
+            PPX_LOG_FATAL(flagName << " KnobStrDropdown: too few allowed choices");
+        }
+        mChoices  = choices;
+        mCallback = callback;
+        auto res  = SetIndex(defaultIndex, true);
+        if (res != SUCCESS)
+            PPX_LOG_FATAL(flagName << "cannot be initialized");
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetIndex(mDefaultIndex); }
+
+    // FlagText format:
+    // --flag_name <"choice1"|"choice2"|...> : flag description
+    std::string FlagText() const override;
+
+    // Updates mIndex and mDefaultIndex, expected flag format:
+    // --flag_name <str>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    int         GetIndex() const { return mIndex; }
+    std::string GetStr() const { return mChoices[mIndex]; }
+
+    // Used for when mIndex needs to be updated outside of UI
+    // Returns success if newI is valid, fails otherwise
+    // If newI is valid:
+    // - If updateDefault, mDefaultIndex is updated
+    // - If newI is different from mIndex:
+    //   - mIndex is updated
+    //   - mCallback is triggered
+    Result SetIndex(int newI, bool updateDefault = false);
+
+    // Used for setting from flags, will set mIndex to appropriate value if newVal matches a choice exactly
+    // Updates mDefaultIndex, mIndex, and calls mCallback with the same logic as above
+    Result SetIndex(std::string newVal, bool updateDefault = false);
+
+private:
+    // mIndex indicates which of the mChoices is selected
+    int                      mIndex;
+    int                      mDefaultIndex;
+    std::vector<std::string> mChoices;
+
+    // mCallback will be called on the new index every time the dropdown is deselected in the UI, or updated with SetIndex()
+    std::function<void(int)> mCallback;
+};
+
+// KnobManager owns the knobs in an application
+class KnobManager
+{
+public:
+    KnobManager() {}
+    virtual ~KnobManager();
+    KnobManager(const KnobManager&)            = delete;
+    KnobManager& operator=(const KnobManager&) = delete;
+
+    // Utilities
+    bool IsEmpty() { return mRoots.empty(); }
+    void ResetAllToDefault();
+
+    // Create knobs
+    template <typename T, typename... ArgsT>
+    T* CreateKnob(Knob* parent, ArgsT... args)
+    {
+        T* knobPtr = new T(std::forward<ArgsT>(args)...);
+
+        // Reject invalid flagName
+        std::string flagName = knobPtr->GetFlagName();
+        for (auto& ch : flagName) {
+            if (isspace(ch)) {
+                PPX_LOG_ERROR("\"" << flagName << "\" cannot be created as it has whitespace");
+                return nullptr;
+            }
+        }
+        if (mFlagNames.count(flagName) > 0) {
+            PPX_LOG_ERROR(flagName << " already exists and cannot be defined again");
```

Comment:
Perhaps this should be an assert since we define all knobs in the application code?